### ISSUE TITLE
Model Enhancement for Hybrid Quantum-classical Neural Networks in the Qiskit Textbook

### ISSUE THREAD

**Issue by cyx617** [qiskit-community/qiskit-textbook#659]
## Is your feature request related to a problem? Please describe.

[The chapter 4.1.5 of the Qiskit textbook](https://qiskit.org/textbook/ch-machine-learning/machine-learning-qiskit-pytorch.html) shows a very intriguing example of hybrid quantum-classical neural networks with PyTorch and Qiskit and it demonstrates the ability of integrating Qiskit and PyTorch in the field of quantum machine learning. But some details of the code might need to be reconsidered. For example, when creating the hybrid neural network, we use the code below
```
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 32, kernel_size=5)
        self.conv2 = nn.Conv2d(32, 64, kernel_size=5)
        self.dropout = nn.Dropout2d()
        self.fc1 = nn.Linear(256, 64)
        self.fc2 = nn.Linear(64, 1)
        self.hybrid = Hybrid(qiskit.Aer.get_backend('qasm_simulator'), 100, np.pi / 2)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.relu(self.conv2(x))
        x = F.max_pool2d(x, 2)
        x = self.dropout(x)
        x = x.view(-1, 256)
        x = F.relu(self.fc1(x))
        x = self.fc2(x)
        x = self.hybrid(x)
        return torch.cat((x, 1 - x), -1)

```
If you print out the shape of the torch tensor in each layer of the neural network, you would see that after the code ```x = x.view(-1, 256)``` the dimensions of the input tensors for fc1 and fc2 layer would be ```(25,256)``` and ```(25,64)``` respectively. In PyTorch, the input for the fc layer (i.e. linear layer) should be represented as the tensor  ```[batch_size, in_features]``` (as describe in [this documentation](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html#torch.nn.Linear)). So the input tensor ```(25,256)``` for fc1 layer means we have a batch size of 25, but we actually set the batch size as 1 in both ```train_loader``` and ```test_loader```. This is a bit confusing. Furthermore, as the fc layer in PyTorch applies the linear transformation only onto the last dimension (i.e. ```in_features```) of the input tensor, the output of the fc2 layer would be of size ```(25,1)```. This tensor is then fed into ```Hybrid``` module and finally goes to the ```HybridFunction``` which defines the forward pass computation and backward pass computation.
```
class HybridFunction(Function):
    """ Hybrid quantum - classical function definition """
    
    @staticmethod
    def forward(ctx, input, quantum_circuit, shift):
        """ Forward pass computation """
        ctx.shift = shift
        ctx.quantum_circuit = quantum_circuit

        expectation_z = ctx.quantum_circuit.run(input[0].tolist())
        result = torch.tensor([expectation_z])
        ctx.save_for_backward(input, result)

        return result
        
    @staticmethod
    def backward(ctx, grad_output):
        """ Backward pass computation """
        input, expectation_z = ctx.saved_tensors
        input_list = np.array(input.tolist())
        
        shift_right = input_list + np.ones(input_list.shape) * ctx.shift
        shift_left = input_list - np.ones(input_list.shape) * ctx.shift
        
        gradients = []
        for i in range(len(input_list)):
            expectation_right = ctx.quantum_circuit.run(shift_right[i])
            expectation_left  = ctx.quantum_circuit.run(shift_left[i])
            
            gradient = torch.tensor([expectation_right]) - torch.tensor([expectation_left])
            gradients.append(gradient)
        gradients = np.array([gradients]).T
        return torch.tensor([gradients]).float() * grad_output.float(), None, None

class Hybrid(nn.Module):
    """ Hybrid quantum - classical layer definition """
    
    def __init__(self, backend, shots, shift):
        super(Hybrid, self).__init__()
        self.quantum_circuit = QuantumCircuit(1, backend, shots)
        self.shift = shift
        
    def forward(self, input):
        return HybridFunction.apply(input, self.quantum_circuit, self.shift)

```
There seems to be an inconsistency of number of parameters fed into the quantum circuit as only the first element of the input tensor ```(25,1)``` is used for the forward pass computation while all of 25 elements are used to calculate the gradient during the backward pass. This might explain the long training time (model training took around 245s per epoch when I run the model on Qiskit text book page) and the fluctuating loss curve of our model as shown in the Qiskit text book. 

## Describe the solution you'd like

The problem is due to the incorrect dimension of the input tensor for fc1 layer and it can be fixed by replacing the code ```x = x.view(-1, 256)``` by ```x = x.view(1,-1)``` where -1 would infer the last dimension as 25* 256. However, considering the resulting large weight matrix (i.e. 64 by 25*256 matrix), this would add much more parameters to the model and hence increase the risk of overfitting as well as longer training time.  So I actually modified the convolutional layers so that we can have a model of moderate size. Here is the code of model architecture
```
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 6, kernel_size=5)
        self.conv2 = nn.Conv2d(6, 16, kernel_size=5)
        self.dropout = nn.Dropout2d()
        self.fc1 = nn.Linear(256, 64)
        self.fc2 = nn.Linear(64, 1)
        self.hybrid = Hybrid(qiskit.Aer.get_backend('qasm_simulator'), 100, np.pi / 2)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.max_pool2d(x, 2)
        x = F.relu(self.conv2(x))
        x = F.max_pool2d(x, 2)
        x = self.dropout(x)
        x = x.view(1, -1)
        x = F.relu(self.fc1(x))
        x = self.fc2(x)
        x = self.hybrid(x)
        return torch.cat((x, 1 - x), -1)

```
where I reduce the number of filters/kernels of each convolutional layer to ensure an input tensor of moderate size for the fc1 layer. I did not change other parts of the code in the example. I run both the current and new model in the jupyter notebook from IBM Quantum Experience. It turned out that the new model achieved a better performance in terms of both train/test loss and running time, as shown in the following table

Model | Train Loss | Test Loss | Training Time |
------------ | ------------- |------------- |------------- |
Current | -0.8937 |-0.8734 |1480s |
New | -0.9871 | -0.9794  | 145s |

In addition, the training process of the new model is more stable, as shown in the training loss curve below

![image](https://user-images.githubusercontent.com/69285126/89703854-577c6480-d981-11ea-8899-280366c4ce41.png)

which might help build confidence for readers who would like to work in the field of quantum machine learning.

The current model in the example can actually runs successfully, but some details might confuse some readers, especially those who don't have too much experience in machine learning or deep learning. So I am not sure if it is necessary to update the code in chapter 4.1.5 of the Qiskit text book. But I hope my report could help with the documentation enhancement.

## Additional context

Since the original example works only for binary classification, I tried to generalized the model to multi-class classification (i.e. problem of classifying instances into one of three or more classes ). Let me take six-class classification for example. Here is the code for the new model architecture

```
class Net1(nn.Module):
    def __init__(self):
        super(Net1, self).__init__()
        self.conv1 = nn.Conv2d(1, 6, kernel_size=5)
        self.conv2 = nn.Conv2d(6, 16, kernel_size=5)
        self.dropout = nn.Dropout2d()
        self.fc1 = nn.Linear(256, 64)
        self.fc2 = nn.Linear(64, 2)
        self.hybrid1 = Hybrid(qiskit.Aer.get_backend('qasm_simulator'), 100, np.pi / 2)
        self.hybrid2 = Hybrid(qiskit.Aer.get_backend('qasm_simulator'), 100, np.pi / 2)
        self.fc3 = nn.Linear(2, 6)

    def forward(self, x):
        x = F.relu(self.conv1(x))
        x = F.max_pool2d(x, 2)
        x = F.relu(self.conv2(x))
        x = F.max_pool2d(x, 2)
        x = self.dropout(x)
        x = x.view(1, -1)
        x = F.relu(self.fc1(x))
        x = self.fc2(x)
        x1 = self.hybrid1(x[0][0].reshape(1,-1))
        x2 = self.hybrid2(x[0][1].reshape(1,-1))
        x = torch.cat((x1,x2),-1).float()
        x = self.fc3(x)
        return x

```
The dimension of the output tensor for fc2 layer has changed from ```(1,1))``` to ```(1,2)``` which means there would be two neurons after fc2 layer. Two hybrid layers are then applied to these two neurons respectively and their corresponding output tensors are concatenated along the last dimension. Since we consider six-class classification in this example, the resulting tensor is then transformed into a tensor of dimension ```(1,6)``` using fc3 layer and the output is fed into the softmax cross entropy loss function. I also modify the code for model training so that the model performance can be monitored during the training process. Here is the code

```
def train(model, opimizer, loss_func, train_loader):
    model.train()

    total_loss = []
    for batch_idx, (data, target) in enumerate(train_loader):
        optimizer.zero_grad()
        # Forward pass
        output = model(data)
        # Calculating loss
        loss = loss_func(output, target)
        # Backward pass
        loss.backward()
        # Optimize the weights
        optimizer.step()

        total_loss.append(loss.item())
    #loss_list.append(sum(total_loss)/len(total_loss))

    return sum(total_loss)/len(total_loss)

def validation(model, test_loader):
    model.eval()
    with torch.no_grad():
        total_loss = []
        correct = 0
        for batch_idx, (data, target) in enumerate(test_loader):
            output = model(data)
            pred = output.argmax(dim=1, keepdim=True) 
            correct += pred.eq(target.view_as(pred)).sum().item()
            loss = loss_func(output, target)
            total_loss.append(loss.item())
    loss = sum(total_loss) / len(total_loss)
    acc = correct / len(test_loader) * 100

    return loss, acc

model = Net1()
optimizer = optim.Adam(model.parameters(), lr=0.001)
loss_func = nn.CrossEntropyLoss()
train_loss_list = []
test_loss_list = []
acc_list = []
epochs = 40
for epoch in range(epochs):
    train_loss = train(model,optimizer,loss_func,train_loader)
    train_loss_list.append(train_loss)
    test_loss, acc = validation(model,test_loader)
    acc_list.append(acc)
    test_loss_list.append(test_loss) 
    print('Training [{:.0f}%]\ttrain_Loss: {:.4f}\ttest_acc: {:.1f}%'.format(
        100. * (epoch + 1) / epochs, train_loss,acc))
    

```
where I define for convenience the ```train``` and ```validation``` functions for model training and model testing respectively. By running this code, we can see the model performance (e.g., loss, accuracy) on the test dataset for every epoch during the training, which is a conventional way to check the model generalization ability. The number of epoch has been doubled as we have more model parameters and training samples than the case for binary classification. After 40 training epochs, the test accuracy of the model can converge to up to **98%** which is good considering the model size and the complexity of the task. Here are a few examples of classification results:

![image](https://user-images.githubusercontent.com/69285126/90091835-ff1add80-dd59-11ea-8513-9f2634ffe3da.png)

In the case of binary classification, the hybrid layer with parameterized quantum circuits  serves as the binary classifier with the output x and 1-x representing the probability of class 1 and 0 respectively. But in the case of multi-class classification, the hybrid layer actually plays the role of an activation function with the output in range (0,1). So I was curious about the difference of performance between this type of activation function and other classical activation functions (e.g. sigmoid and tanh) which also have outputs in range (0,1). Then I modified the model by replacing hybrid functions with sigmoid and tanh functions separately and trained the two new models under the same condition.
It turns out that these two new models have much worse performances than the one with hybrid function, as shown in the images below

![image](https://user-images.githubusercontent.com/69285126/89995828-a0ebed00-dcbc-11ea-900c-305d7ba16517.png)
![image](https://user-images.githubusercontent.com/69285126/89995872-af3a0900-dcbc-11ea-9e47-0d9a7482a37c.png)

My first guessing is this is perhaps due to the vanishing gradient problem which always makes it difficult to train neural networks with gradient based methods (e.g. back propagation). So I checked the gradients of these three activation functions calculated at the early stage of the training (e.g., 10th epoch) and made three histograms respectively for comparison.

![image](https://user-images.githubusercontent.com/69285126/90091637-80be3b80-dd59-11ea-80e6-07f8e9791e88.png)

It turns out that this is indeed the case. As we can see, most of gradients of sigmoid and tanh functions become zero or close to zero at this stage while hybrid function still has a lot of gradients with large values. This can explain why both test loss and accuracy almost have not changed since around 10th epoch. The vanishing gradients of sigmoid and tanh functions can be understood by looking at their function curves

![image](https://user-images.githubusercontent.com/69285126/90014433-a146b100-dcd9-11ea-81db-a0fb0cdfb066.png)

They both have a very small active or unsaturated region, namely the region with larger gradients. So it is easy for them to kill the gradients of the neural network. On the other hand, the hybrid function can be actually calculated analytically in light of [this paper](https://arxiv.org/pdf/1905.13311.pdf). In our case, the Hermitian generator **G** and the observable **A** are both Pauli matrix **Z**, and the initial sate is **|0>**. After the computation, the hybrid function defined as the first formula in the paper can be expressed in terms of sine or cosine functions which are periodic functions and have an infinite number of active regions. So it is much more likely for the hybrid function to avoid the vanishing gradient problem. 

![image](https://user-images.githubusercontent.com/69285126/90089938-494d9000-dd55-11ea-88c0-c94556e4c9fc.png)

It can be seen from the above example that the parameterized quantum circuits might be considered as a good option for activation functions in the neural network models. However, we need further investigations on that as this is just a very simple example. I hope this additional context could trigger more interest in applying quantum circuits to neural networks by leveraging Qiskit.

**Comment by frankharkins** [comment]
@cyx617 thanks for investigating this. I believe @amyami187 wrote the chapter and would be the best person to look at this.

**Comment by amyami187** [comment]
@cyx617 this is a very nice extension! We are aiming to add an additional chapter on barren plateaus as part of the quantum machine learning section of the textbook. I think this would be great in there, where can elaborate on the barren plateaus phenomenon in more detail, generalize the neural network example to multi-class and demonstrate that it falls into this vanishing gradient problem.

**Comment by amyami187** [comment]
@BryceFuller what do you think? :)

**Comment by cyx617** [comment]
@frankharkins Thank you for letting me know that :blush:.

**Comment by cyx617** [comment]
@amyami187 Thank you for your comments :smiley:. I'm very happy that this context could help with the new additional chapter of the qiskit textbook and looking forward to see that chapter in the future. Please let me know if there is anything regarding this work I could help with.

**Comment by BryceFuller** [comment]
This looks fantastic! @amyami187 @cyx617 
The points about the existing implementation having inconsistent input dimensions should be integrated into the existing textbook entry and I agree with Amira the other content would fit well in a chapter on Barren Plateaus.

**Comment by cyx617** [comment]
@BryceFuller Thanks for your comments 😃. Please let me know if there is anything I could help with.

**Comment by frankharkins** [comment]
@cyx617 if you like, you can update the article with your corrections and make a pull request.

**Comment by cyx617** [comment]
@frankharkins No problem. I'll make a pull request regarding the corrections. Thanks for your suggestion.

**Comment by cyx617** [comment]
Issue fixed
